debian/docs/sb_dbexpimp.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.40.10.
.TH SB_DBEXPIMP "1" "June 2014" "sb_dbexpimp" "User Commands"
.SH NAME
sb_dbexpimp.py \- Bayes database export/import
.SH DESCRIPTION
This utility has the primary function of exporting and importing
a spambayes database into/from a CSV file.  This is useful in a number
of scenarios.
.IP
Platform portability of database \- CSV files can be exported and
imported across platforms (Windows and Linux, for example).
.IP
Database implementation changes \- databases can survive database
implementation upgrades or new database implementations.  For example,
if a dbm implementation changes between python x.y and python x.y+1...
.IP
Database reorganization \- an export followed by an import reorgs an
existing database, <theoretically> improving performance, at least in
some database implementations.
.IP
Database sharing \- it is possible to distribute particular databases
for research purposes, database sharing purposes, or for new users to
have a 'seed' database to start with.
.IP
Database merging \- multiple databases can be merged into one quite
easily by specifying \fB\-m\fR on an import.  This will add the two database
nham and nspams together and for wordinfo conflicts, will add spamcount
and hamcount together.
.SS "Usage:"
.IP
sb_dbexpimp [options]
.IP
options:
.TP
\fB\-e\fR
: export
.TP
\fB\-i\fR
: import
.IP
\fB\-f\fR: FN : flat file to export to or import from
\fB\-p\fR: FN : name of pickled database file to use
\fB\-d\fR: FN : name of dbm database file to use
\fB\-m\fR     : merge import into an existing database file.  This is
.IP
meaningful only for import. If omitted, a new database
file will be created.  If specified, the imported
wordinfo will be merged into an existing database.
Run dbExpImp \fB\-h\fR for more information.
.IP
\fB\-o\fR: section:option:value :
.IP
set [section, option] in the options database to value
.TP
\fB\-h\fR
: help
.PP
If neither \fB\-p\fR nor \fB\-d\fR is specified, then the values in your configuration
file (or failing that, the defaults) will be used.  In this way, you may
convert to and from storage formats other than pickle and dbm.
.PP

.IP
Export pickled mybayes.db into mybayes.db.export as a CSV file
.IP
sb_dbexpimp \fB\-e\fR \fB\-p\fR mybayes.db \fB\-f\fR mybayes.db.export
.IP
Import mybayes.db.export into a new DBM mybayes.db
.IP
sb_dbexpimp \fB\-i\fR \fB\-d\fR mybayes.db \fB\-f\fR mybayes.db.export
.IP
Convert a bayes database from pickle to DBM
.IP
sb_dbexpimp \fB\-e\fR \fB\-p\fR abayes.db \fB\-f\fR abayes.export
sb_dbexpimp \fB\-i\fR \fB\-d\fR abayes.db \fB\-f\fR abayes.export
.IP
Create a new DBM database (newbayes.db) from two
.IP
DBM databases (abayes.db, bbayes.db)
sb_dbexpimp \fB\-e\fR \fB\-d\fR abayes.db \fB\-f\fR abayes.export
sb_dbexpimp \fB\-e\fR \fB\-d\fR bbayes.db \fB\-f\fR bbayes.export
sb_dbexpimp \fB\-i\fR \fB\-d\fR newbayes.db \fB\-f\fR abayes.export
sb_dbexpimp \fB\-i\fR \fB\-m\fR \fB\-d\fR newbayes.db \fB\-f\fR bbayes.export
.SH "SEE ALSO"
The full documentation for
.B option
is maintained as a Texinfo manual.  If the
.B info
and
.B option
programs are properly installed at your site, the command
.IP
.B info option
.PP
should give you access to the complete manual.


CHANGELOG.txt
[Note that all dates are in English, not American format - i.e. day/month/year]

Release 1.0.4
=============
Tony Meyer        09/03/2005  Use a blocking socket when connecting with sb_server, so that with Python 2.4 we don't end up with never-ending reports of errors.
Tony Meyer        07/02/2005  sb_server: When non-verbose, only report unique connection errors once per hour.

Release 1.0.3
=============
Tony Meyer        28/01/2005  Distribute mscvr71.dll, as Microsoft recommend.
Tony Meyer        28/01/2005  The 'fix' for escaping subject lines was incorrect; do it correctly.

Release 1.0.2
=============
Tony Meyer        14/01/2005  Try and work around the reasonably common problems with imaplib and OS X.
Tony Meyer        04/01/2005  If we are reloading the options, optionsPathname will already be set, so if we are  a windows user using the default pathname, the object won't be reloaded.  Fix so a reload is always done.
Tony Meyer        02/01/2005  UserInterface.py: cgi.escape the configuration filename
Tony Meyer        22/12/2004  Make Dibbler.py compatible with Python 2.2.1.
Tony Meyer        17/12/2004  message.py: Change the deprecated function to work in Python 2.4.
Tony Meyer        06/12/2004  Fix [ 1078923 ] Unicode support incomplete
Tony Meyer        06/12/2004  Fix the regex that the auth digest used, and handle the auth digest responses tha IE 6.0 and Firefox 1.0 give.
Tony Meyer        29/11/2004  Subject lines are not cgi.escape()d in the web interface, which might cause errors - fixed.

Release 1.0.1
=============
Tony Meyer        15/11/2004  Fix a bug in sb_dbexpimp.py where merging into an existing dbm file might lose training data.
Tony Meyer        15/11/2004  sb_dbexpimp.py: Fail if the csv file doesn't exist that we are trying to import from rather than  keeping going, which made no sense.
Tony Meyer        11/11/2004  The installer wasn't offered to install a startup items shortcut, so fix that.  This is a non-ideal patch, but appears to be the only way Inno will work.
Tony Meyer        03/11/2004  Fix [ 1022848 ] sb_dbexpimp.py crashes while importing into pickle file
Tony Meyer        03/11/2004  Fix [ 831864 ] sb_mboxtrain.py: flock vs. lockf
Tony Meyer        03/11/2004  Fix [ 922063 ] Intermittent sb_filter.py failure with URL pickle
Tony Meyer        29/09/2004  Fix [ 1036601 ] typo on advanced config web page
Tony Meyer        28/10/2004  Add [ 715248 ] Pickle classifier should save to a temp file first
Tony Meyer        21/10/2004  Fix [ 1051081 ] uncaught socket timeoutexception slurping URLs
Tony Meyer        18/10/2004  TestDriver: If show_histograms was False, then the global ham/spam histogram never had the stats computed, but this gets used later, so the script would die with an AtrributeError. Fix that.
Tony Meyer        13/10/2004  Fix mySQL storage option for the case where the server does not support rollbacks.
Sjoerd Mullender  02/10/2004  imapfilter: Quote the search string that tries to find the message again that was just saved.
Tony Meyer        30/09/2004  Fix [ 903905 ] IMAP Configuration Error
Tony Meyer        15/09/2004  sb_upload: Clarify docstring so that it's more clear what this script does. The -n / --null command line option didn't actually do anything; change it so that it does.
Tony Meyer        23/07/2004  For proxy handler for version checking, the proxy port needs to be an integer, not a string.
Tony Meyer        19/07/2004  Fix [ 990700 ] Changes to asyncore in Python 2.4 break ServerLineReader
Kenny Pitt        17/07/2004  Fix [941639] and [986353].  Use a non-standard extension for our py2exe created zip to get around Windows extensions that automatically expand zip files.
Tony Meyer        14/07/2004  Fix [ 790757 ] signal handler created with wrong # of args
Tony Meyer        14/07/2004  Fix [ 944109 ] notate_to/subject option valid values should be dynamic
Tony Meyer        14/07/2004  Fix [ 959937 ] "Invalid server" message not always correct
Skip Montanaro    10/07/2004  tte.py: 2.3 compatibility: add reversed() function
Tony Meyer        09/07/2004  Update test_storage.py test to reflect (current) correct way to call open_storage. Fixes part of [ 981970 ] tests failing.
Tony Meyer        09/07/2004  Using -u with sb_server had been broken.  Fix this.

1.0 Final
=========
(no changes from 1.0rc2)

Release Candidate 2
===================
Tony Meyer        10/06/2004  Install license with binaries.
Mark Hammond      01/06/2004  Fix Tony's fix for [ 943397 ] Cannot choose more than one Inbox folder to filter
Tony Meyer        02/06/2004  Fix [ 961713 ] server error when trying to change Statistics Options
Skip Montanaro    31/05/2004  Compatcsv: Abandon regexes as bad way to parse csv file - add simple unit test as well
Skip Montanaro    31/05/2004  sb_dbexpimp.py: handle case of importing the old object craft csv module.
Skip Montanaro    22/05/2004  sb_dbexpimp.py: Add a fake UnicodeDecodeError for Python 2.2 compatibility.
Skip Montanaro    22/05/2004  Compatcsv: correct a list chomping bug (forgot a slicing ":") and use .readline() instead of .next() for Python 2.2 compatibility
Tony Meyer        18/05/2004  Fix [ 955442 ] default_bayes_customize.ini not handled right
Tony Meyer        14/05/2004  message.py: If we trigger the TypeError problem, try to do our best to put the message together anyway.
Tony Meyer        14/05/2004  Handle people using a storage type that isn't pickle/dbm (like one of the sql ones).
Tony Meyer        13/05/2004  If a user tried to send a bug report before entering in any servers to proxy, this would generate an IndexError.  Fix that.
Tony Meyer        12/05/2004  Fix [ 943397 ] Cannot choose more than one Inbox folder to filter

Release Candidate 1
===================
Skip Montanaro    05/05/2004  sb_filter, sb_bnfilter use new mboxutils as_string().

scripts/sb_dbexpimp.py
#! /usr/bin/env python

"""sb_dbexpimp.py - Bayes database export/import

Classes:


Abstract:

    This utility has the primary function of exporting and importing
    a spambayes database into/from a flat file.  This is useful in a number
    of scenarios.

    Platform portability of database - flat files can be exported and
    imported across platforms (winduhs and linux, for example)

    Database implementation changes - databases can survive database
    implementation upgrades or new database implementations.  For example,
    if a dbm implementation changes between python x.y and python x.y+1...

    Database reorganization - an export followed by an import reorgs an
    existing database, <theoretically> improving performance, at least in
    some database implementations

    Database sharing - it is possible to distribute particular databases
    for research purposes, database sharing purposes, or for new users to
    have a 'seed' database to start with.

    Database merging - multiple databases can be merged into one quite
    easily by specifying -m on an import.  This will add the two database
    nham and nspams together (assuming the two databases do not share
    corpora) and for wordinfo conflicts, will add spamcount and hamcount
    together.

    Spambayes software release migration - an export can be executed before
    a release upgrade, as part of the installation script.  Then, after the
    new software is installed, an import can be executed, which will
    effectively preserve existing training.  This eliminates the need for
    retraining every time a release is installed.

    Others?  I'm sure I haven't thought of everything...

Usage:
    sb_dbexpimp [options]

        options:
            -e     : export
            -i     : import
            -v     : verbose mode (some additional diagnostic messages)
            -f: FN : flat file to export to or import from
            -p: FN : name of pickled database file to use
            -d: FN : name of dbm database file to use
            -m     : merge import into an existing database file.  This is
                     meaningful only for import. If omitted, a new database
                     file will be created.  If specified, the imported
                     wordinfo will be merged into an existing database.
                     Run dbExpImp -h for more information.
            -o: section:option:value :
                     set [section, option] in the options database to value

            -h     : help

Examples:

    Export pickled mybayes.db into mybayes.db.export as a csv flat file
        sb_dbexpimp -e -p mybayes.db -f mybayes.db.export

    Import mybayes.db.export into a new DBM mybayes.db
        sb_dbexpimp -i -d mybayes.db -f mybayes.db.export

    Convert a bayes database from pickle to DBM
        sb_dbexpimp -e -p abayes.db -f abayes.export
        sb_dbexpimp -i -d abayes.db -f abayes.export

    Create a new DBM database (newbayes.db) from two
        DBM databases (abayes.db, bbayes.db)
        sb_dbexpimp -e -d abayes.db -f abayes.export
        sb_dbexpimp -e -d bbayes.db -f bbayes.export
        sb_dbexpimp -i -d newbayes.db -f abayes.export
        sb_dbexpimp -i -m -d newbayes.db -f bbayes.export

To Do:
    o Suggestions?

"""

# This module is part of the spambayes project, which is Copyright 2002
# The Python Software Foundation and is covered by the Python Software
# Foundation license.

from __future__ import generators

__author__ = "Tim Stone <tim@fourstonesExpressions.com>"


# Python 2.2 compatibility stuff
try:
    True, False
except NameError:
    True, False = 1, 0

try:
    import csv
    # might get the old object craft csv module - has no reader attr 
    if not hasattr(csv, "reader"): 
        raise ImportError 
except ImportError:
    import spambayes.compatcsv as csv

try:
    x = UnicodeDecodeError
except NameError:
    UnicodeDecodeError = UnicodeError
else:
    del x


import spambayes.storage
from spambayes.Options import options
import sys, os, getopt, errno, re
import urllib
from types import UnicodeType

def uquote(s):
    if isinstance(s, UnicodeType):
        s = s.encode('utf-8')
    return s

# Heaven only knows what encoding non-ASCII stuff will be in
# Try a few common western encodings and punt if they all fail
def uunquote(s):
    for encoding in ("utf-8", "cp1252", "iso-8859-1"):
        try:
            return unicode(s, encoding)
        except UnicodeDecodeError:
            pass
    # punt
    return s

def runExport(dbFN, useDBM, outFN):
    bayes = spambayes.storage.open_storage(dbFN, useDBM)
    if useDBM == "dbm":
        words = bayes.db.keys()
        words.remove(bayes.statekey)
    else:
        words = bayes.wordinfo.keys()

    try:
        fp = open(outFN, 'wb')
    except IOError, e:
        if e.errno != errno.ENOENT:
            raise

    writer = csv.writer(fp)

    nham = bayes.nham;
    nspam = bayes.nspam;

    print "Exporting database %s to file %s" % (dbFN, outFN)
    print "Database has %s ham, %s spam, and %s words" \
            % (nham, nspam, len(words))

    writer.writerow([nham, nspam])

    for word in words:
        wi = bayes._wordinfoget(word)
        hamcount = wi.hamcount
        spamcount = wi.spamcount
        word = uquote(word)
        writer.writerow([word, hamcount, spamcount])

def runImport(dbFN, useDBM, newDBM, inFN):

    if newDBM:
        try:
            os.unlink(dbFN)
        except OSError:
            pass

        try:
            os.unlink(dbFN+".dat")
        except OSError:
            pass

        try:
            os.unlink(dbFN+".dir")
        except OSError:
            pass

    bayes = spambayes.storage.open_storage(dbFN, useDBM)

    fp = open(inFN, 'rb')
    rdr = csv.reader(fp)
    (nham, nspam) = rdr.next()

    if newDBM:
        bayes.nham = int(nham)
        bayes.nspam = int(nspam)
    else:
        bayes.nham += int(nham)
        bayes.nspam += int(nspam)

    if newDBM:
        impType = "Importing"
    else:
        impType = "Merging"

    print "%s file %s into database %s" % (impType, inFN, dbFN)

    for (word, hamcount, spamcount) in rdr:
        word = uunquote(word)

        # Can't use wordinfo[word] here, because wordinfo
        # is only a cache with dbm!  Need to use _wordinfoget instead.
        wi = bayes._wordinfoget(word)
        if wi is None:
            wi = bayes.WordInfoClass()

        wi.hamcount += int(hamcount)
        wi.spamcount += int(spamcount)

        bayes._wordinfoset(word, wi)

    print "Storing database, please be patient.  Even moderately sized"
    print "databases may take a very long time to store."
    bayes.store()
    print "Finished storing database"

    if useDBM == "dbm" or useDBM == True:
        words = bayes.db.keys()
        words.remove(bayes.statekey)
    else:
        words = bayes.wordinfo.keys()

    print "Database has %s ham, %s spam, and %s words" \
           % (bayes.nham, bayes.nspam, len(words))


if __name__ == '__main__':

    try:
        opts, args = getopt.getopt(sys.argv[1:], 'iehmvd:p:f:o:')
    except getopt.error, msg:
        print >>sys.stderr, str(msg) + '\n\n' + __doc__
        sys.exit()

    useDBM = "pickle"
    newDBM = True
    dbFN = None
    flatFN = None
    exp = False
    imp = False

    for opt, arg in opts:
        if opt == '-h':
            print >>sys.stderr, __doc__
            sys.exit()
        elif opt == '-f':
            flatFN = arg
        elif opt == '-e':
            exp = True
        elif opt == '-i':
            imp = True
        elif opt == '-m':
            newDBM = False
        elif opt == '-v':
            options["globals", "verbose"] = True
        elif opt in ('-o', '--option'):
            options.set_from_cmdline(arg, sys.stderr)
    dbFN, useDBM = spambayes.storage.database_type(opts)

    if (dbFN and flatFN):
        if exp:
            runExport(dbFN, useDBM, flatFN)
        if imp:
            runImport(dbFN, useDBM, newDBM, flatFN)
    else:
        print >>sys.stderr, __doc__
